src/main/java/org/apache/ibatis/executor/resultset/PendingConstructorCreation.java

```diff
@@ -0,0 +1,232 @@
+ */
+final class PendingConstructorCreation {
+
+  private final Class<?> resultType;
+  private final List<Class<?>> constructorArgTypes;
+  private final List<Object> constructorArgs;
+  private final Map<Integer, PendingCreationMetaInfo> linkedCollectionMetaInfo;
+  private final Map<String, Collection<Object>> linkedCollectionsByResultMapId;
+  private final Map<String, List<PendingConstructorCreation>> linkedCreationsByResultMapId;
+
+  PendingConstructorCreation(Class<?> resultType, List<Class<?>> types, List<Object> args) {
+    // since all our keys are based on result map id, we know we will never go over args size
+    final int maxSize = types.size();
+    this.linkedCollectionMetaInfo = new HashMap<>(maxSize);
+    this.linkedCollectionsByResultMapId = new HashMap<>(maxSize);
+    this.linkedCreationsByResultMapId = new HashMap<>(maxSize);
+    this.resultType = resultType;
+    this.constructorArgTypes = types;
+    this.constructorArgs = args;
+  }
+
+  @SuppressWarnings("unchecked")
+  Collection<Object> initializeCollectionForResultMapping(ObjectFactory objectFactory, ResultMap resultMap,
+      ResultMapping constructorMapping, Integer index) {
+    final Class<?> parameterType = constructorMapping.getJavaType();
+    if (!objectFactory.isCollection(parameterType)) {
+      throw new ExecutorException(
+          "Cannot add a collection result to non-collection based resultMapping: " + constructorMapping);
+    }
+
+    final String resultMapId = constructorMapping.getNestedResultMapId();
+    return linkedCollectionsByResultMapId.computeIfAbsent(resultMapId, (k) -> {
+      // this will allow us to verify the types of the collection before creating the final object
+      linkedCollectionMetaInfo.put(index, new PendingCreationMetaInfo(resultMap.getType(), resultMapId));
+
+      // will be checked before we finally create the object) as we cannot reliably do that here
+      return (Collection<Object>) objectFactory.create(parameterType);
+    });
+  }
+
+  void linkCreation(ResultMap nestedResultMap, PendingConstructorCreation pcc) {
+    final String resultMapId = nestedResultMap.getId();
+    final List<PendingConstructorCreation> pendingConstructorCreations = linkedCreationsByResultMapId
+        .computeIfAbsent(resultMapId, (k) -> new ArrayList<>());
+
+    if (pendingConstructorCreations.contains(pcc)) {
+      throw new ExecutorException("Cannot link inner pcc with same value, MyBatis programming error!");
+    }
+
+    pendingConstructorCreations.add(pcc);
+  }
+
+  void linkCollectionValue(ResultMapping constructorMapping, Object value) {
+    // not necessary to add null results to the collection (is this a config flag?)
+    if (value == null) {
+      return;
+    }
+
+    final String resultMapId = constructorMapping.getNestedResultMapId();
+    if (!linkedCollectionsByResultMapId.containsKey(resultMapId)) {
+      throw new ExecutorException("Cannot link collection value for resultMapping: " + constructorMapping
```

Comment:
Should we standardize the type of exceptions this code can throw? (internal vs user error)

Reply:
`ExecutorException` looks fine.
If you feel strongly about this, we can reassess it later.